docs/USAGE-GUIDE.md

```diff
@@ -500,6 +500,208 @@ legacy SHA-1 algorithms in CertificateVerify messages if TLS1.2 has been negotia
+or received all data.
+* In non-blocking mode an s2n-tls I/O function may return while there is
+still I/O pending. In this case `s2n_error_get_type()` will report **S2N_ERR_T_BLOCKED**
+and the value of the s2n_blocked_status parameter will be set appropriately. s2n-tls I/O functions
+should be called repeatedly until the **blocked** parameter is **S2N_NOT_BLOCKED**.
+
+Servers in particular usually prefer non-blocking mode. In blocking mode, a single connection
+blocks the thread while waiting for more IO. In non-blocking mode, multiple connections
+can make progress by returning control while waiting for more IO.
+
+To use s2n-tls in non-blocking mode, set the underlying file descriptors as non-blocking. 
+For example:
+```c
+int flags = fcntl(fd, F_GETFL, 0);
+if (flags < 0) return S2N_FAILURE;
+if (fcntl(fd, F_SETFL, flags | O_NONBLOCK) < 0) return S2N_FAILURE;
+```
+
+### Errors and Alerts
+
+If the peer sends an alert, the next call to a read IO method will report **S2N_FAILURE** and
+`s2n_error_get_type()` will return **S2N_ERR_T_ALERT**. The specific alert received
+is available by calling `s2n_connection_get_alert()`. By default, s2n-tls treats
+all alerts as fatal.
+
+If s2n-tls encounters a fatal error, the next call to a write IO method will send
+a close_notify alert to the peer. Except for a few exceptions, s2n-tls does not
+send specific alerts in order to avoid leaking information that could be used for
+a sidechannel attack. To ensure that the alert is sent, `s2n_shutdown()` should
+be called after an error.
+
+### Performing the TLS Handshake
+
+Before application data can be sent or received, an application must perform a handshake
+to establish a TLS connection with the peer.
+
+To perform the handshake, call `s2n_negotiate()` until it either returns **S2N_SUCCESS**
+or returns **S2N_FAILURE** without a **S2N_ERR_T_BLOCKED** error.
+
+### Application Data
+
+After the TLS handshake, an application can send and receive encrypted data.
+
+Although most s2n-tls APIs are not thread-safe, `s2n_send()` and `s2n_recv()`
+may be called simultaneously from different threads.
+
+Even if an application only intends to send data or only intends to receive data,
+it should implement both send and receive in order to handle alerts and post-handshake
+TLS control messages like session tickets.
+
+#### Sending Application Data
+
+`s2n_send()` and its variants encrypt and send application data to the peer.
+The sending methods return the number of bytes written and may indicate a partial
+write, even if blocking IO is used.
+
+Unlike OpenSSL, repeated calls to `s2n_send()` should not duplicate the original
+parameters, but should update the inputs per the indication of size written.
+
+For example:
+```c
+char data[10]; /* Some data we want to write */
+
+s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+int written = 0;
+do {
+    int w = s2n_send(conn, data + written, sizeof(data) - written, &blocked);
```

Comment:
This does a `0` length write at the end. Is that ok? I'm assuming that doesn't close the connection or anything?

Reply:
It's okay, the last write will just be a no-op.

But we should be able to do the loop by just checking written<sizeof(data), without using blocked. I'm going to update the example to do that and see if it's any better.

Reply:
A 0-length write would be fine, that's just a no-op.

But would it do a 0-length write? Wouldn't the last non-zero write return >0 and not set blocked?